AdminSelfUpgrade.php

```diff
@@ -1998,13 +2038,55 @@ public function ajaxProcessCleanDatabase()
 		$this->nextQuickInfo[] = $this->l('The database has been cleaned.');
 	}
 
+	public function getServerFullBaseUrl()
+	{
+		$s = $_SERVER;
+		$ssl = (!empty($s['HTTPS']) && $s['HTTPS'] == 'on');
+		$sp = strtolower($s['SERVER_PROTOCOL']);
+		$protocol = substr($sp, 0, strpos($sp, '/')) . (($ssl) ? 's' : '');
+		$port = $s['SERVER_PORT'];
+		$port = ((!$ssl && $port == '80') || ($ssl && $port == '443')) ? '' : ':' . $port;
+		$host = isset($s['HTTP_HOST']) ? $s['HTTP_HOST'] : null;
+		$host = isset($host) ? $host : $s['SERVER_NAME'] . $port;
+		return $protocol . '://' . $host;
+	}
+
 	/**
-	 * This function now replaces doUpgrade.php or upgrade.php
+	 * Implement the upgrade based on upgrade.php from the downloaded archive
+	 *
+	 * @return bool
 	 *
-	 * @return void
 	 */
-	public function doUpgrade()
+	public function doUpgrade17()
 	{
+		if (!defined('__PS_BASE_URI__')) {
+			define('__PS_BASE_URI__', realpath(dirname($_SERVER['SCRIPT_NAME'])) . '/../../');
+		}
+		$hostName = $this->getServerFullBaseUrl();
+		$url = $hostName .
+			'/'.$this->install_autoupgrade_dir.'/upgrade/upgrade.php?autoupgrade=1&deactivateCustomModule=' .
+			(int)$this->deactivateCustomModule.'&updateDefaultTheme='.(int)$this->updateDefaultTheme.
+			'&keepMails='.(int)$this->keepMails.'&changeToDefaultTheme='.(int)$this->changeToDefaultTheme.
+			'&adminDir='.base64_encode($this->adminDir).'&idEmployee='.(int)$_COOKIE['id_employee'];
+		$json = Tools14::file_get_contents($url);
+		$result = json_decode($json, true);
```

Comment:
there's no max exec  time in the core upgrade file, but when it fails, the popup asking something when wrong appear and propose to restore the backup